### PR TITLE
DHIS2-434

### PR DESCRIPTION
**Pull request by stian-sandvold**

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/reservedvalue/ReservedValue.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        return Objects.equals( ownerObject, that.ownerObject ) &&
+            Objects.equals( ownerUid, that.ownerUid ) &&
+            Objects.equals( key, that.key ) &&
+            Objects.equals( value, that.value );
+    }
+
+    @Override
+    public int hashCode()
+    {
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/reservedvalue/ReservedValue.java]: remove empty line

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/textpattern/BaseMethodType.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package org.hisp.dhis.textpattern;
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/textpattern/BaseMethodType.java]: Copyright

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/textpattern/MethodType.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package org.hisp.dhis.textpattern;
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/textpattern/MethodType.java]: Copyright

File: dhis-2/dhis-api/src/test/java/org/hisp/dhis/textpattern/TestTextPatternMethod.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
+package org.hisp.dhis.textpattern;
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/test/java/org/hisp/dhis/textpattern/TestTextPatternMethod.java]: Copyright

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/DefaultReservedValueService.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        List<String> resultList = new ArrayList<>();
+
+        TextPatternSegment generatedSegment = getGeneratedSegment( textPattern );
+
+        String key = textPatternService.resolvePattern( textPattern, values );
+
+        // Used for searching value tables
+        String valueKey = (generatedSegment != null ? key.replaceAll( generatedSegment.getRawSegment(), "%" ) : key);
+
+        ReservedValue reservedValue = new ReservedValue( textPattern.getOwnerObject().name(), textPattern.getOwnerUID(), key,
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/DefaultReservedValueService.java]: codestyle

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/hibernate/HibernateReservedValueStore.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    @Override
+    public boolean useReservedValue( String ownerUID, String value )
+    {
+        return getQuery( "DELETE FROM ReservedValue WHERE owneruid = :uid AND value = :value" )
+            .setParameter( "uid", ownerUID )
+            .setParameter( "value", value )
+            .executeUpdate() == 1;
+    }
+
+    // Helper methods:
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/hibernate/HibernateReservedValueStore.java]: Adjust style of "Helper methods". Not consistent with others. Maybe we should have a style convention for this?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/textpattern/TextPatternParser.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    );
+
+    /**
+     * Parses an expression, identifying segments and builds an IDExpression.
+     * throws exception if syntax is invalid
+     *
+     * @param pattern the expression to parse
+     * @return IDExpression representing the expression
+     */
+    public static TextPattern parse( String pattern )
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/textpattern/TextPatternParser.java]: Long method. Maybe divide into smaller methods to get a better understanding of what is going on. As well as removing some of the comments.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/reservedvalue/DefaultReservedValueServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    @Autowired
+    private DefaultReservedValueService reservedValueService;
+
+    @Autowired
+    private ReservedValueStore reservedValueStore;
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Autowired
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/reservedvalue/DefaultReservedValueServiceTest.java]: Moved dep up?

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/reservedvalue/hibernate/HibernateSequentialNumberCounterStoreTest.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
+package org.hisp.dhis.reservedvalue.hibernate;
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/reservedvalue/hibernate/HibernateSequentialNumberCounterStoreTest.java]: Copyright

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/textpattern/TestDefaultTextPatternService.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
+package org.hisp.dhis.textpattern;
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/textpattern/TestDefaultTextPatternService.java]: Copyright

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/textpattern/TestTextPatternParser.java [hunk 11]
```diff
@@ -0,0 +1,216 @@
+package org.hisp.dhis.textpattern;
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/textpattern/TestTextPatternParser.java]: Copyright

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/textpattern/TestTextPatternParser.java [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+    {
+
+        testParseOK( EXAMPLE_TEXT_SEGMENT, TextPatternMethod.TEXT );
+    }
+
+    @Test
+    public void testParseTextWithEscapedQuotes()
+        throws TextPatternParser.TextPatternParsingException
+    {
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/textpattern/TestTextPatternParser.java]: remove emptyline

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/TrackedEntityAttributeController.java [hunk 13]
```diff
@@ -56,47 +64,141 @@
...
     }
 
-    @RequestMapping( value = "/{id}/generate", method = RequestMethod.GET, produces = { ContextUtils.CONTENT_TYPE_JSON, ContextUtils.CONTENT_TYPE_JAVASCRIPT } )
-    public @ResponseBody String queryTrackedEntityInstancesJson(
-        @PathVariable String id,
-        Model model,
-        HttpServletResponse response ) throws Exception
+    // Helpers
+
+    private List<String> reserve( String id, int numberToReserve, int daysToLive )
```
Comment (henninghaakonsen) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/TrackedEntityAttributeController.java]: Can we move this to a service class?

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/reservedvalue/DefaultReservedValueServiceTest.java (hunk 8 above)
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/reservedvalue/DefaultReservedValueServiceTest.java]: removed :)

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/textpattern/TextPatternParser.java (hunk 7 above)
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/textpattern/TextPatternParser.java]: Thanks, will clean up this class later :)

**Pull request merged**